src/compressor/lz4/LZ4Compressor.cc

```diff
@@ -121,16 +121,12 @@ int LZ4Compressor::decompress(ceph::buffer::list::const_iterator &p,
   LZ4_streamDecode_t lz4_stream_decode;
   LZ4_setStreamDecode(&lz4_stream_decode, nullptr, 0);
 
-  ceph::buffer::ptr cur_ptr = p.get_current_ptr();
-  ceph::buffer::ptr *ptr = &cur_ptr;
-  std::optional<ceph::buffer::ptr> data_holder;
-  if (compressed_len != cur_ptr.length()) {
-    data_holder.emplace(compressed_len);
-    p.copy_deep(compressed_len, *data_holder);
-    ptr = &*data_holder;
-  }
-
-  char *c_in = ptr->c_str();
+  ceph::buffer::list indata;
+  // this does a shallow copy
+  p.copy(compressed_len, indata);
+  // if the input isn't fragmented, c_str() costs almost nothing.
+  // otherwise rectifying copy will be taken
+  const char* c_in = indata.c_str();
```

Comment:
I'm concerned that this is going to result in more memcpy's than the previous implementation, not for messenger where everything should be in a single buffer anyway, but for other users of decompression such as BlueStore.

For example looking at BlueStore::_prepare_read_ioc it creates a multi-buffer bufferlist with one buffer for each compressed blob that it reads. Later in BlueStore::_generate_read_result_bl it calls BlueStore::_decompress which calls cp->decompress passing this bufferlist.

In the old code if you have a multi-buffer bufferlist and just try to decompress data within one buffer segment of the list there is no memcpy. With this change there will be a memcpy because indata.c_str() has no knowledge about which sub-set of the bufferlist will be accessed.

Reply:
Previously there was no mempcy only when the rest the underlying ptr was `compressed_len` in size.
Currently there is no memcpy in this case as well.

Reply:
Sorry, missed that the copy of p.copy(compressed_len,indata) is only going to put one buffer into indata if the data to be decompressed is all in a single buffer ptr.